tests/integration_tests/functional/test_cpu_features.py

```diff
@@ -142,7 +144,100 @@ def test_brand_string(test_microvm_with_api, network_config):
+    """
+    Test MSRs that are available to the Guest.
+
+    This test boots a Firecracker uVM and tries to read a set of MSRs from the guest.
+    The Guest MSR list is compared against a list of MSRs that are expected when running
+    on a particular host kernel and with a particular Guest CPU template.
+    The list is different for each kernel version because Firecracker relies on
+    MSR emulation provided by KVM. If KVM emulation changes, then the MSR list
+    available to the guest might change also.
+    The list is also dependant on CPUID (CPU templates) since some MSRs are not available
+    if CPUID features are disabled.
+    Lastly, this tests also checks for MSR values against the baseline. This helps validate
+    that defaults have not changed due to emulation implementation changes in the kernel.
+
+    @type: functional
+    """
+    test_microvm = test_microvm_with_api_and_msrtools
+    test_microvm.spawn()
+    test_microvm.basic_config(vcpu_count=1)
+    # Set template as specified in the `cpu_template` parameter.
+    response = test_microvm.machine_cfg.put(
+        vcpu_count=1,
+        mem_size_mib=256,
+        cpu_template=cpu_template,
+    )
+    assert test_microvm.api_session.is_status_no_content(response.status_code)
+    _, _, _ = test_microvm.ssh_network_config(network_config, "1")
+    response = test_microvm.actions.put(action_type="InstanceStart")
+    assert test_microvm.api_session.is_good_response(response.status_code)
+
+    ssh_connection = net_tools.SSHConnection(test_microvm.ssh_config)
+    ssh_connection.scp_file(
+        "../resources/tests/msr/msr_reader.sh", "/bin/msr_reader.sh"
+    )
+    _, stdout, stderr = ssh_connection.execute_command("/bin/msr_reader.sh")
+    assert stderr.read() == ""
+
+    # Load results read from the microvm
+    microvm_df = pd.DataFrame(pd.read_csv(stdout))
+
+    # Load baseline
+    baseline_csv = pd.read_csv(
+        f"../resources/tests/msr/msr_list_{cpu_template}_{utils.get_kernel_version(level=1)}.csv"
+    )
+    baseline_df = pd.DataFrame(baseline_csv)
+
+    # We first want to see if the same set of MSRs are exposed in the microvm.
+    # Drop the VALUE columns and compare the 2 dataframes.
+    impl_diff = pd.concat(
+        [microvm_df.drop(columns="VALUE"), baseline_df.drop(columns="VALUE")],
+        keys=["microvm", "baseline"],
+    ).drop_duplicates(keep=False)
+    assert impl_diff.empty, f"\n {impl_diff}"
+
+    # Now drop the STATUS column to compare values for each MSR
+    microvm_val_df = microvm_df.drop(columns="STATUS")
+    baseline_val_df = baseline_df.drop(columns="STATUS")
+
+    # Some MSR values should not be checked since they can change at Guest runtime.
+    # Also some MSRs are different based on Guest configuration and kernel used.
+    # FS and GS change on task switch and arch_prctl.
+    # TSC is obviously different for each Guest.
+    ignore_msrs = ["0x10", "0x6e0", "0xc0000100", "0xc0000101"]
```

Comment:
Would there be a nice/appropriate place to define these in msr_index.rs with some minimal description of what these MSRs can do?

Reply:
Not sure what would be a different place to put this list. I'm open for suggestions.
As for documentation, I've added the comment above. `0x10` is `MSR_IA32_TSC`, `0x6e0` is `MSR_IA32_TSCDEADLINE`, `0xc0000100` is `MSR_FS_BASE`, `0xc0000101` is `MSR_GS_BASE`.
I can add comments for each but the definitions are in `msr_index.rs` or `msr_index.h` in the kernel.